File: usercorn/arch/arm/linux.py
from unicorn.arm_const import *
import os
import sys

def syscall(cls):
    # regs = [X86_REG_EAX, X86_REG_EBX, X86_REG_ECX, X86_REG_EDX, X86_REG_ESI, X86_REG_EDI, X86_REG_EBP]
    # num, a1, a2, a3, a4, a5, a6 = [cls.reg_read(r) for r in regs]
    ret = 0
    if num == -1:
        pass
    else:
        print 'Unsupported syscall:', num
        sys.exit(1)
    # cls.reg_write(UC_X86_REG_EAX, ret)

def interrupt(cls, intno):
    if intno == 0:
        syscall(cls)
    else:
        raise NotImplementedError('unhandled interrupt %d' % intno)

File: go/arch/sparc/linux.go
package sparc

import (
	"fmt"
	"github.com/lunixbochs/ghostrace/ghost/sys/num"
	uc "github.com/unicorn-engine/unicorn/bindings/go/unicorn"

	co "github.com/lunixbochs/usercorn/go/kernel/common"
	"github.com/lunixbochs/usercorn/go/kernel/linux"
	"github.com/lunixbochs/usercorn/go/models"
)

// TODO: sparc linux syscall abi regs
var LinuxRegs = []int{}

type LinuxKernel struct {
	*linux.LinuxKernel
}

func LinuxKernels(u models.Usercorn) []interface{} {
	kernel := &LinuxKernel{linux.NewKernel()}
	return []interface{}{kernel}
}

func LinuxSyscall(u models.Usercorn) {
	// TODO: handle errors or something
	g1, _ := u.RegRead(uc.SPARC_REG_G1)
	// TODO: add sparc x86 syscall numbers to ghostrace
	name, _ := num.Linux_x86[int(g1)]
	ret, _ := u.Syscall(int(g1), name, co.RegArgs(u, LinuxRegs))
	u.RegWrite(uc.SPARC_REG_O0, ret)
}

// TODO: add sparc syscall convention support
func LinuxInterrupt(u models.Usercorn, intno uint32) {
	panic(fmt.Sprintf("unknown interrupt: %d", intno))
}

func init() {
	Arch.RegisterOS(&models.OS{
		Name:      "linux",
		Kernels:   LinuxKernels,
		Init:      linux.StackInit,
		Interrupt: LinuxInterrupt,
	})
}
